Cross updating with Microservices eBook v2.1.02 - Chapter 9 (#8504)
dotnet/docs: docs/standard/microservices-architecture/secure-net-microservices-web-applications/azure-key-vault-protects-secrets.md

---
title: Using Azure Key Vault to protect secrets at production time

> description: Securing .NET Microservices and Web Applications | Azure Key Vault is an excellent way to handle application secrets that are completely controlled by administrators. Administrators can even assign and revoke development values without developers having to handle them. Take a look here.

Reviewer: Securing => Secure

Reviewer: Remove "Take a look here"

Reviewer 2: No piped sentences in the description.

Author: Done.

author: mjrousos
ms.author: wiwagn
ms.date: 10/19/2018
---
# Use Azure Key Vault to protect secrets at production time

Secrets stored as environment variables or stored by the Secret Manager tool are still stored locally and unencrypted on the machine. A more secure option for storing secrets is [Azure Key Vault](https://azure.microsoft.com/services/key-vault/), which provides a secure, central location for storing keys and secrets.

The **Microsoft.Extensions.Configuration.AzureKeyVault** package allows an ASP.NET Core application to read configuration information from Azure Key Vault. To start using secrets from an Azure Key Vault, you follow these steps:

> 1. Register your application as an Azure AD application. (Access to key vaults is managed by Azure AD.) This can be done through the Azure management portal.\

Reviewer 2: ```suggestion
1. Register your application as an Azure AD application. (Access to key vaults is managed by Azure AD.) This can be done through the Azure management portal.
```

   Alternatively, if you want your application to authenticate using a certificate instead of a password or client secret, you can use the [New-AzureRmADApplication](https://docs.microsoft.com/powershell/module/azurerm.resources/new-azurermadapplication) PowerShell cmdlet. The certificate that you register with Azure Key Vault needs only your public key. (Your application will use the private key.)

2. Give the registered application access to the key vault by creating a new service principal. You can do this using the following PowerShell commands:

   ```powershell
   $sp = New-AzureRmADServicePrincipal -ApplicationId "<Application ID guid>"
   Set-AzureRmKeyVaultAccessPolicy -VaultName "<VaultName>" -ServicePrincipalName $sp.ServicePrincipalNames[0] -PermissionsToSecrets all -ResourceGroupName "<KeyVault Resource Group>"
   ```

> 3. Include the key vault as a configuration source in your application by calling the `IConfigurationBuilder.AddAzureKeyVault` extension method when you create an `IConfigurationRoot` instance. Note that calling `AddAzureKeyVault` will require the application ID that was registered and given access to the key vault in the previous steps.

Reviewer: Prefer xref for .NET APIs

Author: Done.

> 3. Include the key vault as a configuration source in your application by calling the <xref:Microsoft.Extensions.Configuration.AzureKeyVaultConfigurationExtensions.AddAzureKeyVault> extension method when you create an <xref:Microsoft.Extensions.Configuration.IConfigurationRoot > instance. Note that calling `AddAzureKeyVault` will require the application ID that was registered and given access to the key vault in the previous steps.

Reviewer 2: ```suggestion
3. Include the key vault as a configuration source in your application by calling the <xref:Microsoft.Extensions.Configuration.AzureKeyVaultConfigurationExtensions.AddAzureKeyVault%2A?displayProperty=nameWithType> extension method when you create an <xref:Microsoft.Extensions.Configuration.IConfigurationRoot> instance. Note that calling `AddAzureKeyVault` requires the application ID that was registered and given access to the key vault in the previous steps.
```

   You can also use an overload of `AddAzureKeyVault` that takes a certificate in place of the client secret by just including a reference to the [Microsoft.IdentityModel.Clients.ActiveDirectory](https://www.nuget.org/packages/Microsoft.IdentityModel.Clients.ActiveDirectory) package.

> **It's a best practice to register Azure Key Vault as the last configuration provider so it can  override configuration values from previous providers.**

Reviewer: Instead of bold, I'd consider the "Important" tag from our [style guide](https://github.com/dotnet/docs/blob/master/styleguide/template.md#alerts)

Author: done.

> >[!IMPORTANT] It's a best practice to register Azure Key Vault as the last configuration provider so it can  override configuration values from previous providers.

Reviewer 2: ```suggestion
> [!IMPORTANT]
> We recommend you to register Azure Key Vault as the last configuration provider, so it can override configuration values from previous providers.
```
